translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,388 @@
+// Copyright 2019 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opencensus
+
+import (
+	"strings"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+func ocToInternal(td consumerdata.TraceData) data.TraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := data.NewResourceSpans(resource, nil)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+
+	if spanCount != 0 {
+		// Create slice that holds pointers and another slice that holds structs at once
+		// to avoid individual allocations of structs.
+		spans := data.NewSpanSlice(spanCount)
+		spanPtrs := make([]*data.Span, 0, spanCount)
+
+		for i, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			// Point one element in slice of pointers to an element in slice of structs.
+			destSpan := &spans[i]
+
+			ocSpanToInternal(destSpan, ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
```

Comment:
Can be that more than one span have the same different Resource?

Reply:
It is not possible for data received from network. Possible for locally created data, but I don't think we need to optimize for it. It is not mandatory anyway, separate Resource objects work too.